### PR TITLE
Tests: Switch to PHPUnit

### PR DESCRIPTION
**Pull request by drupol**
This PR:

* [x] Switch to [PHPUnit](https://www.phpunit.de) framework and get rid of [PHPSpec](http://phpspec.net/).
* [x] Use a contrib plugin for PHPUnit to compare iterables: https://github.com/loophp/phpunit-iterable-assertions 
* [x] Depends on #228 because the PHPUnit plugin needs a `ClosureIterator` so I created [`loophp/iterators`](https://github.com/loophp/iterators) to use it in `loophp/collection` and `loophp/phpunit-iterable-assertions`.
* [x] Add a `phpunit.xml` configuration file
* [x] Use PHP 7.4 and 8.1 on continuous integration 
* [x] Fix issue with infection, see https://github.com/infection/infection/issues/1615
* [x] Investigate why Infection score is different on PHP 7.4 and 8.1.

This is a huge work - I'm testing the best way to implement tests. So far I tested:
- [x] One file per test - One dataprovider per file
- [x] One big file containing all the dataproviders and one single test

I really like having one single basic tests and multiple dataproviders... if you have a better idea, let me know.

Follows https://github.com/loophp/collection/discussions/173

### CODE REVIEW

File: tests/unit/Operation/AllCollectionTest.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ */
+final class AllCollectionTest extends CollectionTestCase
+{
+    public function dataProvider()
+    {
+        yield [
+            [1, 2, 3],
+            [1, 2, 3],
+            false,
+        ];
```
Comment (AlexandruGG) [inline tests/unit/Operation/AllCollectionTest.php:25]:
```suggestion
    /**
    * @return Generator<array{0: array<mixed>, 1: array<mixed>, 2: bool}>
    */
    public function dataProvider(): Generator
    {
        yield [
            [1, 2, 3],
            [1, 2, 3],
            false,
        ];
```

the types are missing here but I think it's because the test paths are ignored in static analysis. I would say since we're doing this from scratch again it would be helpful to analyse the test namespace as well. Tests effectively show how the code is being used by users, so static analysis will help spot any issues from that point of view as well

Comment (drupol) [reply tests/unit/Operation/AllCollectionTest.php:25]: Oh! should we test the types in PHPUnit ?

Comment (sanmai) [reply tests/unit/Operation/AllCollectionTest.php]: That's the thing to do if your life feel too easy :sweat_smile: 
Seriously, these things hold no value.

File: .github/workflows/mutation-tests.yml [hunk 2]
```diff
@@ -33,7 +33,7 @@ jobs:
               uses: shivammathur/setup-php@2.16.0
               with:
                   php-version: ${{ matrix.php-versions }}
-                  extensions: gd,mbstring,pcov
+                  extensions: gd,mbstring,pcov,xdebug
```
Comment (sanmai) [inline .github/workflows/mutation-tests.yml:36]: Shouldn't need xdebug here.

File: tests/unit/Operation/GenericCollectionTest.php [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+declare(strict_types=1);
+
+namespace tests\loophp\collection\Operation;
+
+use loophp\collection\Collection;
+use tests\loophp\collection\CollectionTestCase;
+
+/**
+ * @internal
+ * @covers \loophp\collection\Collection
```
Comment (sanmai) [inline tests/unit/Operation/GenericCollectionTest.php:17]:
```suggestion
```
This is why there are so many uncovered mutants.

**Comment by sanmai** [comment]
THB I don't understand what's going on here. That GrumPHP thing might be at fault. If I run Infection like so:
```
vendor/bin/infection -j4
```
I have no issues whatsoever.

**Comment by drupol** [comment]
> THB I don't understand what's going on here. That GrumPHP thing might be at fault. If I run Infection like so:

Try: `./vendor/bin/grumphp --tasks=infection -vvv` to get the command it runs:

`Command: '/home/pol/dev/git/collection/vendor/bin/infection' '--no-interaction' '--ignore-msi-with-no-mutations' '--threads=4' '--test-framework=phpunit' '--configuration=infection.json' '--min-msi=90' '--min-covered-msi=90'
`

**Comment by sanmai** [comment]
Ah, looks like you need to raise the timeout:
> 611 mutants required more time than configured

File: grumphp.yml [hunk 4]
```diff
@@ -61,7 +61,7 @@ parameters:
             clover_file: build/logs/clover.xml
             level: 99
         infection:
-            threads: 10
+            threads: 4
```
Comment (sanmai) [inline grumphp.yml:64]: You tests are using very little RAM, keeping this at 10 threads should work as it seems.

Comment (drupol) [reply grumphp.yml:64]: Ok, good to know! I mostly do that so my laptop doesn't get stuck when running the tests :)

**Comment by drupol** [comment]
Thanks for helping @sanmai ! Going to close the issue I opened with the solution.

File: grumphp.yml [hunk 5]
```diff
@@ -1,66 +1,68 @@
...
+        phpunit:
+            always_execute: true
+        clover_coverage:
+            clover_file: build/logs/clover.xml
+            level: 99
+        infection:
+            threads: 4
+            test_framework: phpunit
+            configuration: infection.json
+            min_msi: 90
```
Comment (sanmai) [inline grumphp.yml:67]:
```suggestion
            min_msi: 60
```

File: grumphp.yml [hunk 6]
```diff
@@ -1,66 +1,68 @@
...
+            min_covered_msi: 90
```
Comment (sanmai) [inline grumphp.yml:68]:
```suggestion
            min_covered_msi: 85
```

File: grumphp.yml (hunk 5 above)
Comment (drupol) [reply grumphp.yml:67]: When the migration will be over, we'll be at least at 90%... there are many tests that I need to migrate.

It's a huge work, and I'm doing that slowly, piece by piece :)

Comment (sanmai) [reply grumphp.yml:67]: Sure thing! It might be easier to move this thing out of your radar until everything else is done, and then you can raise the numbers to any value you want.

Comment (drupol) [reply grumphp.yml:67]: It's ok... I'll take my time, no worries about that. I guess we should come back >90% by the end of the week.

Comment (sanmai) [reply grumphp.yml:67]: Best of luck!

Comment (drupol) [reply grumphp.yml:67]: Thanks for the help and piece of advice!

Comment (drupol) [reply grumphp.yml:67]: Finally ! I'm above 90% :) except for PHP 7.4.

File: tests/unit/Operation/AllCollectionTest.php (hunk 1 above)
Comment (AlexandruGG) [reply tests/unit/Operation/AllCollectionTest.php]: Hmm I think the value comes from getting the types right on the source code itself 😄. If you look at the closed PRs you will see what I mean :).

Anyway we can add that in a separate PR, so don't worry worry about it for now @drupol

File: phpunit.xml [hunk 7]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd"
+         bootstrap="./vendor/autoload.php"
+         colors="true"
+         executionOrder="random"
+         failOnWarning="true"
+         failOnRisky="true"
+         verbose="true"
```
Comment (AlexandruGG) [inline phpunit.xml:9]:
```suggestion
         executionOrder="random"
         beStrictAboutChangesToGlobalState="true"
         beStrictAboutOutputDuringTests="true"
         beStrictAboutCoversAnnotation="true"
         forceCoversAnnotation="true"
         failOnRisky="true"
         failOnEmptyTestSuite="true"
         verbose="true"
```

a few more config parameters I think we should use, you can see what they do here: https://phpunit.readthedocs.io/en/9.5/configuration.html#the-phpunit-element

File: tests/unit/CollectionSpecificOperationTest.php [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+use stdClass;
+use tests\loophp\collection\Traits\AssertIdenticalIterable;
+use tests\loophp\collection\Traits\GenericCollectionProviders;
+use function gettype;
+
+/**
+ * @internal
+ * @coversDefaultClass \loophp\collection
+ */
+final class CollectionSpecificOperationTest extends TestCase
```
Comment (AlexandruGG) [inline tests/unit/CollectionSpecificOperationTest.php:33]: Is this one going to have tests with each of the operations or is it more for specific ones that we feel need specific tests?

Comment (drupol) [reply tests/unit/CollectionSpecificOperationTest.php:33]: Indeed, only the more specific one. I'm very bad at naming, do you have a better suggestion ?

Comment (AlexandruGG) [reply tests/unit/CollectionSpecificOperationTest.php:33]: The name is makes sense I think, so it's all good

**Comment by drupol** [comment]
@AlexandruGG I'm thinking about getting rid of Infection tests on PHP 7.4 and keep only on PHP 8.1. WDYT?

**Comment by AlexandruGG** [comment]
> @AlexandruGG I'm thinking about getting rid of Infection tests on PHP 7.4 and keep only on PHP 8.1. WDYT?

Are we sure the 8.1 ones are working properly? Can't really tell from the output which mutants are generated and killed. Seems a bit odd that more mutants would be generated and not killed on 7.4 🤔

**Comment by drupol** [comment]
Migration is finally over, that was tough.

The only remaining step is to understand why the infection score is completely different on PHP 7.4 and PHP 8.1.

Maybe @sanmai can help on this? Do you have some clue?

**Comment by sanmai** [comment]
I wish I had, but I can't see what Infection finds on PHP 7.4.

**Comment by drupol** [comment]
Let me run the command manually on CI.

**Comment by drupol** [comment]
Here's the logs:
* PHP 8.1: https://github.com/loophp/collection/runs/4604629898?check_suite_focus=true
* PHP 7.4: https://github.com/loophp/collection/runs/4604629846?check_suite_focus=true

Check the "Run Grumphp" step.

**Comment by sanmai** [comment]
Here's what I had on PHP 7.4:

```
Processing source code files: 242/242
.: killed, M: escaped, U: uncovered, E: fatal error, X: syntax error, T: timed out, S: skipped, I: ignored

UUUUUU............................................   ( 50 / 865)
........M.M....M.......MM...M....M..M....M........   (100 / 865)
........M....................MM...................   (150 / 865)
...MM.MM......................EE..E..E.......E..M.   (200 / 865)
...........................M......................   (250 / 865)
..M.MMMMM..............................M...MM.....   (300 / 865)
......M...........................................   (350 / 865)
............MM..M................M................   (400 / 865)
.MMMM.M......M.........M.......EEE................   (450 / 865)
..............................M...MM..............   (500 / 865)
.....MMMMM............MMM......MM.MMMM..M.MM.MMMMM   (550 / 865)
..M....M................................MM.......M   (600 / 865)
....................M..MM.......................MM   (650 / 865)
M.......................................M.M...MM.M   (700 / 865)
.........................MM.MMM.M.........M.......   (750 / 865)
..................................................   (800 / 865)
....M............M.....................M..........   (850 / 865)
...........M.MM                                      (865 / 865)

865 mutations were generated:
     758 mutants were killed
       0 mutants were configured to be ignored
       6 mutants were not covered by tests
      93 covered mutants were not detected
       8 errors were encountered
       0 syntax errors were encountered
       0 time outs were encountered
       0 mutants required more time than configured

Metrics:
         Mutation Score Indicator (MSI): 88%
         Mutation Code Coverage: 99%
         Covered Code MSI: 89%

Please note that some mutants will inevitably be harmless (i.e. false positives).

Time: 2m 39s. Memory: 0.07GB
```

And here's for PHP 8:

```
Processing source code files: 242/242
.: killed, M: escaped, U: uncovered, E: fatal error, X: syntax error, T: timed out, S: skipped, I: ignored

UUUUUUUUU.........................................   ( 50 / 865)
..................MM....M.......MM...M..M....M...M   (100 / 865)
..................................................   (150 / 865)
..................................................   (200 / 865)
..................................................   (250 / 865)
..................................................   (300 / 865)
..................................................   (350 / 865)
..................................................   (400 / 865)
..................................................   (450 / 865)
.....................................M..M.M.......   (500 / 865)
..................................................   (550 / 865)
..................................................   (600 / 865)
........................MM........................   (650 / 865)
..................................................   (700 / 865)
..............................................M...   (750 / 865)
..................................................   (800 / 865)
..........M...............................M.......   (850 / 865)
..............T                                      (865 / 865)

865 mutations were generated:
     838 mutants were killed
       0 mutants were configured to be ignored
       9 mutants were not covered by tests
      17 covered mutants were not detected
       0 errors were encountered
       0 syntax errors were encountered
       1 time outs were encountered
       0 mutants required more time than configured

Metrics:
         Mutation Score Indicator (MSI): 96%
         Mutation Code Coverage: 98%
         Covered Code MSI: 98%

Please note that some mutants will inevitably be harmless (i.e. false positives).

Time: 55s. Memory: 0.07GB
```
Infection finds the same number of mutants on PHP 8 (that's expected for PHP 7.4 project), but more mutants escape on PHP 7.4: looks like this is because PHP 8.0 is much more strict in terms what's allowed and what's not. If a mutation causes a notice on PHP 7.4, it could be the reason for a fat warning in 8.0, and that's a test failure.

Bottom line: I didn't find anything wrong here.

**Comment by drupol** [comment]
Ok ! I have the confirmation of what I thought ! Thanks !!!

**Comment by drupol** [comment]
Got another feedback from @maks-rafalko on Slack

image

**Comment by sanmai** [comment]
I can second this. It also runs quite a bit faster on PHP 8 :+1:

**Comment by drupol** [comment]
Ok, this PR is done. Ready for review.

File: .gitattributes [hunk 9]
```diff
@@ -20,3 +20,4 @@ phpstan-baseline.neon export-ignore
 phpstan-docs-baseline.neon export-ignore
 phpstan-unsupported-baseline.neon export-ignore
 phpstan.neon.dist export-ignore
+phpunit.xml export-ignore
```
Comment (AlexandruGG) [inline .gitattributes:23]: I guess we can remove the `phpspec.yml` from line 15

Comment (drupol) [reply .gitattributes:23]: Sure, will do it right now.

File: docs/pages/principles.rst [hunk 10]
```diff
@@ -165,6 +165,6 @@ we can use `iterator_to_array`_ to convert this back to a normal array when need
 .. _ClosureIterator: https://github.com/loophp/collection/blob/master/src/Iterator/ClosureIterator.php
 .. _invoke: https://www.php.net/manual/en/language.oop5.magic.php#object.invoke
 .. _iterator_to_array: https://www.php.net/manual/en/function.iterator-to-array.php
-.. _iterateAs: https://www.phpspec.net/en/stable/cookbook/matchers.html#iterateas-matcher
+.. _iterateAs: https://github.com/loophp/phpunit-iterable-assertions
```
Comment (AlexandruGG) [inline docs/pages/principles.rst:168]:
```suggestion
.. _assertIdenticalIterable: https://github.com/loophp/phpunit-iterable-assertions
```

File: tests/unit/CollectionConstructorsTest.php [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+            [2 => 'c', 3 => 'd', 4 => 'e'],
+            Collection::fromIterable($generator)
+        );
+
+        // TODO: Reimplement with PHPUNIT.
+        /*
+        $subject
+            ->shouldThrow(Error::class)
+            ->during('count');
+         */
```
Comment (AlexandruGG) [inline tests/unit/CollectionConstructorsTest.php:195]: did you forget about these or are they for a future PR?

File: .gitattributes (hunk 9 above)
Comment (drupol) [reply .gitattributes]: Done!

File: tests/unit/CollectionGenericOperationTest.php [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+     *
+     * @param mixed $expected
+     */
+    public function testGenericScalar(
+        string $operation,
+        array $parameters,
+        iterable $actual,
+        $expected
+    ): void {
+        self::assertEquals(
```
Comment (AlexandruGG) [inline tests/unit/CollectionGenericOperationTest.php:168]:
```suggestion
        self::assertSame(
```

I would recommend we avoid `assertEquals` altogether, and use `assertSame` for scalars because it's more akin to `===`, checking both the value and the type. For objects we should use [assertObjectEquals](https://phpunit.readthedocs.io/en/9.5/assertions.html#assertobjectequals)

File: tests/unit/Iterator/PsrCacheIteratorTest.php [hunk 13]
```diff
@@ -69,20 +73,12 @@ public function it_can_cache_data(CacheItemPoolInterface $cache): void
             ->hasItem(1)
             ->willReturn(false);
 
-        $this->next();
+        $iterator->next();
 
-        $this
-            ->current()
-            ->shouldReturn('b');
+        self::assertEquals('b', $iterator->current());
```
Comment (AlexandruGG) [inline tests/unit/Iterator/PsrCacheIteratorTest.php:78]:
```suggestion
        self::assertSame('b', $iterator->current());
```

this applies everywhere

File: tests/unit/CollectionConstructorsTest.php (hunk 11 above)
Comment (drupol) [reply tests/unit/CollectionConstructorsTest.php]: Fixed :)

File: tests/unit/Iterator/PsrCacheIteratorTest.php (hunk 13 above)
Comment (drupol) [reply tests/unit/Iterator/PsrCacheIteratorTest.php]: I replaced `assertEquals` with `assertSame` everywhere.

File: tests/unit/Utils/CallbacksArrayReducerTest.php [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+            static fn (int $v): bool => 0 === $v % 2,
+        ];
+
+        $iterator = new ArrayIterator(range(0, 10));
+
+        self::assertTrue(CallbacksArrayReducer::or()($callbacks, 0, 0, $iterator));
+
+        self::assertTrue(CallbacksArrayReducer::or()($callbacks, 13, 0, $iterator));
+
+        self::assertFalse(CallbacksArrayReducer::or()($callbacks, 3, 0, $iterator));
```
Comment (AlexandruGG) [inline tests/unit/Utils/CallbacksArrayReducerTest.php:54]: it's preferable to have a single assertion per test, and the data for these to come from a data provider

**Review by AlexandruGG — APPROVED** [review]
Great work in this @drupol! 🚀

File: tests/unit/Utils/CallbacksArrayReducerTest.php (hunk 14 above)
Comment (drupol) [reply tests/unit/Utils/CallbacksArrayReducerTest.php:54]: That is definitely true. Going to do it now, then merge!

Comment (drupol) [reply tests/unit/Utils/CallbacksArrayReducerTest.php]: It's done, WDYT ?

Comment (drupol) [reply tests/unit/Utils/CallbacksArrayReducerTest.php]: I guess it's ok, going to merge this now so we can move forward ! I'm excited !

**Comment by drupol** [comment]
Pheew ! It's done.
